### ISSUE TITLE
Peak picking ROI vs non-ROI

### ISSUE THREAD

**Issue by michaelwitting** [sneumann/xcms#215]
Hi,

today I realized another issue. I wanted to analyze the MS1 level of some standard measurements.
The data was generated on a Bruker maXis and contains one MS1 trace and three MS2 traces, all with the same isolation mass (Tryptophan in the attached example), but different collision energies (10, 20, 40 eV).
I wanted to pick peaks in the MS1 level to identify in-source fragments, adducts etc.

However, CentWave is not finding the correct [M+H] for Tryptophan (205.1 @ 2.3 min), only if I'm using a ROI-List around that region.

Am I doing something stupid/wrong/whatever or is there a bug?

Below some code and a link to the file.

```
#load libraries
library(xcms)
library(RColorBrewer)

## Use socket based parallel processing on Windows systems
if (.Platform$OS.type == "unix") {
  register(bpstart(MulticoreParam(2)))
} else {
  register(bpstart(SnowParam(2)))
} 

#load util functions (tbl2roi)
source("functions\\utils.R")

###############################################################################################
# Tryptophan
###############################################################################################
## Read the full data at once (all MS level). Advantage: easier to link
## MS1 and MS2 data and file has to be read only once. 
xrms <- readMSData("MSMSpos_0005_RP_Bruker.mzML", mode = "onDisk", centroided = TRUE)

# filter out the MS1 & MS2 level
xrms_ms1 <- filterMsLevel(xrms, msLevel = 1)
bpis <- chromatogram(xrms_ms1, aggregationFun = "max", mz = c(205.0, 205.2))
plot(bpis)

#peak detection with ROI
ms1Tbl <- data.frame(mz = 205.1073,
                     rtMin = 100,
                     rtMax = 200)

ms1Roi <- tbl2ROImzOnly(ms1Tbl, xrms_ms1, 0.5)
cwp <- CentWaveParam(roiList = ms1Roi)
ms1data <- findChromPeaks(xrms_ms1, param = cwp, msLevel = 1)
ms1peakROI <- chromPeaks(ms1data)

#peak detection without ROI
cwp <- CentWaveParam()
ms1data <- findChromPeaks(xrms_ms1, param = cwp, msLevel = 1)
ms1peakNoROI <- chromPeaks(ms1data)
ms1peakNoROI <- ms1peakNoROI[which(abs(ms1peakNoROI[,1] - 205.1) < 0.1),]
```

https://www.dropbox.com/sh/esekicd5749cir5/AADPpJkqKsKhjGhc2ulGQIUda?dl=0

**Comment by jotsetung** [comment]
The default `peakwidth` parameter is wrong for your data. Your peaks are much shorter:
What I usually do is plot the raw data points for some known compounds to get a good guess of centwave parameter values:
```r
library(xcms)
library(RColorBrewer)
xrms <- readMSData("MSMSpos_0005_RP_Bruker.mzML", mode = "onDisk", centroided = TRUE)

					# filter out the MS1 & MS2 level
xrms_ms1 <- filterMsLevel(xrms, msLevel = 1)

## Let's get the complete raw data for the selected slice:
ms_data <- extractMsData(xrms_ms1, mz = c(205.0, 205.2), rt = c(128, 148))[[1]]

#' Create a plot that combines a XIC and a mz/rt 2D plot.
#'
#' @param x `data.frame` such as returned by the `extractMsData` function. Only
#'     a single `data.frame` has to be submitted.
#'
#' @param main `character(1)` specifying the title.
#'
#' @param mfrow `numeric(2)` defining the plot layout. This will be passed
#'     directly to `par(mfrow = mfrow)`. See `par` for more information. Setting
#'     `mfrow = NULL` avoids calling `par(mfrow = mfrow)` hence allowing to
#'     pre-define the plot layout.
#'
#' @author Johannes Rainer
#' 
#' @md
#' 
#' @noRd
plot_msdata <- function(x, main = "", cex = 1, mfrow = c(2, 1)) {
    require(lattice)
    require(RColorBrewer)
    cols <- level.colors(x$i, at = do.breaks(range(x$i), nint = 256),
			 col.regions =
			     colorRampPalette(rev(brewer.pal(9, "YlGnBu"))))
    if (length(mfrow) == 2)
	par(mfrow = mfrow)
    par(mar = c(0, 4, 2, 1))
    x_split <- split(x, f = x$rt)
    ints <- lapply(x_split, function(z) sum(z$i))
    plot(as.numeric(names(ints)), ints, main = main, xlab = "", xaxt = "n",
	 ylab = "", las = 2, pch = 21, bg = cols, col = "grey", cex = cex)
    mtext(side = 4, line = 0, "intensity", cex = par("cex.lab"))
    grid(col = "#00000020")
    par(mar = c(3.5, 4, 0, 1))
    plot(x$rt, x$mz, main = "", pch = 21, bg = cols, col = "grey",
	 xlab = "", ylab = "", yaxt = "n", cex = cex)
    axis(side = 2, las = 2)
    grid(col = "#00000020")
    mtext(side = 1, line = 2.5, "retention time", cex = par("cex.lab"))
    mtext(side = 4, line = 0, "mz", cex = par("cex.lab"))
}

plot_msdata(ms_data)
```

the `plot_msdata` is still a prototype, but it might do it into `xcms` at some point:
![rplot](https://user-images.githubusercontent.com/5506112/29920929-06353784-8e50-11e7-98fe-0eddb5a60ab6.png)

Zooming in as the mz range was quite broad:
```r
## Get the trace for the peak
ms_data <- extractMsData(xrms_ms1, mz = c(205.095, 205.105),
                         rt = c(128, 148))[[1]]
plot_msdata(ms_data)
```
![rplot](https://user-images.githubusercontent.com/5506112/29920968-31956516-8e50-11e7-8f72-186508ee995a.png)

First thing which is clear is that the default `peakwidth` parameter from `xcms` is much too broad. Your peak is about 5 seconds long, so I would set `peakwidth = c(2, 10)`. And that actually does the trick. In addition you could also finetune the `ppm` parameter:
```r
## o `ppm`
##   What would be the ppm (scatter of the measurements around the *real* mz)
##   mz values rt 138-146
mzr <- ms_data[ms_data$rt > 138 & ms_data$rt < 146, "mz"]
##   ppm:
diff(range(mzr)) / mean(mzr) * 1e6
[1] 1.665854
```
which is really tiny - up to you to decide whether you want to lower the `ppm`.

The result with the tuned `peakwidth`:
```r
> cwp <- CentWaveParam(peakwidth = c(2, 10))
> ms1data <- findChromPeaks(xrms_ms1, param = cwp, msLevel = 1)
chromPeaks(ms1data, mz = c(205.09, 205.11))
Detecting mass traces at 25 ppm ... OK
Detecting chromatographic peaks in 4153 regions of interest ... OK: 487 found.
>  
          mz    mzmin    mzmax       rt   rtmin   rtmax     into   intb   maxo
[1,] 205.0968 205.0968 205.0968 141.5598 139.128 144.807 766369.7 766272 234030
       sn sample is_filled
[1,] 7390      1         0
```
So your peak was identified.

**Comment by michaelwitting** [comment]
Oh man... peakwidth was the only parameter I didn't test... Tried sn, ppm, noise...
Well it was stupidity from my side :-D

**Comment by jotsetung** [comment]
no stupidity, it's just always good to look at the complete data.

**Comment by michaelwitting** [comment]
I had Bruker DataAnalysis open in parallel and checked masses etc... But never came to my mind to check peak width. Well if working only with UPLC you some get blind that there could be broader Peaks

**Comment by gmhhope** [comment]
It is my fortune to be able to dig out this issue which has a nice description of finding optimal peak widths. @jorainer. However, for the peak width for the untargeted method, FMHO, we can expect very different peak widths. But the exhaustive search for peak widths is not possible. What can be risky if I set the peak width too wide? e.g., if I find minimal peak widths of common metabolites are around 2.  But I may have several other unknown metabolites that is less than 0.5. Will I miss them because I set peak width minimum = 2, and so can I relax the parameter to even `0`. 

Then the question comes down to - what can be problematic if I set the peak width range too relax?

**Comment by sneumann** [comment]
The min peakwidth is not a hard threshhold, so nothing is immediately discarded. Instead, the peakwidth determines the widths of the wavelets being used. So with peak width minimum = 2 you don't have a wavelet perfectly matching the narrow ones. So a nicely shaped 0.5 peak has a chance to be included. Small, non-symmetric and narrow ones might not. Make sure you have enough scans/sec to resolve a 0.5 peak with more than 1-2 scans. Yours, Steffen
